controller/data/data.go

```diff
@@ -2,18 +2,71 @@ package data
 
 import (
 	"fmt"
+	"strconv"
 	"strings"
+	"time"
 
-	controller "github.com/flynn/flynn/controller/client"
 	ct "github.com/flynn/flynn/controller/types"
 	"github.com/flynn/flynn/pkg/postgres"
 	"github.com/flynn/flynn/pkg/shutdown"
+	"github.com/flynn/flynn/pkg/typeconv"
 	"github.com/inconshreveable/log15"
 )
 
-var ErrNotFound = controller.ErrNotFound
+var ErrNotFound = ct.ErrNotFound
 var logger = log15.New("component", "controller/data")
 
+const DEFAULT_PAGE_SIZE = 1000
+
+type PageToken struct {
+	CursorID *string
+	Size     int
+}
+
+// ParsePageToken decodes a PageToken from a string of the format
+// '<cursorID>|<size>'
+func ParsePageToken(tokenStr string) (*PageToken, error) {
+	token := &PageToken{}
+	if tokenStr == "" {
+		token.Size = DEFAULT_PAGE_SIZE
+		return token, nil
+	}
+	parts := strings.SplitN(tokenStr, "|", 2)
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("error parsing pageToken %q: expected two pipe separated parts, got %d", tokenStr, len(parts))
+	}
+	if parts[0] != "" {
+		token.CursorID = &parts[0]
+	}
+	if parts[1] != "" && parts[1] != "0" {
+		token.Size, _ = strconv.Atoi(parts[1])
+	}
+	if token.Size == 0 {
+		token.Size = DEFAULT_PAGE_SIZE
+	}
+	return token, nil
+}
+
+func (t *PageToken) String() string {
+	if t == nil {
+		return ""
+	}
+	var cursorID string
+	if t.CursorID != nil {
+		cursorID = *t.CursorID
+	}
+	return fmt.Sprintf("%s|%d", cursorID, t.Size)
+}
+
+const cursorIDTimeFormat = "2006-01-02T15:04:05.999999Z07:00"
```

Comment:
Is there a reason why this isn't just microseconds since the unix epoch?

Reply:
No, updated in 56b3067b